lighthouse-core/audits/seo/structured-data-automatic.js

```diff
@@ -0,0 +1,163 @@
+  /* eslint-disable-next-line max-len */
+  description: 'Structured data contains rich metadata about a web page. The data is used in search results and social sharing. Invalid metadata will affect how the page appears in these contexts. This audit currently validates a subset of JSON-LD rules. See also the manual audit below to learn how to validate other types of structured data.',
+  /** Explanatory message stating what percentage of JSON-LD structured data snippets are invalid */
+  displayValue: '{validSnippetProportion, number, percent} valid snippets',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+class StructuredDataAutomatic extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'structured-data-automatic',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['ScriptElements'],
+    };
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @return {Promise<LH.Audit.Product>}
+   */
+  static async audit(artifacts) {
+    const jsonLDElements = artifacts.ScriptElements.filter(
+      script => script.type === 'application/ld+json' && !!script.content);
+
+    if (jsonLDElements.length === 0) {
+      return {
+        notApplicable: true,
+        score: 1,
+      };
+    }
+
+    const validatedSnippets = await Promise.all(jsonLDElements.map(async (element) => {
+      // We don't want to show empty lines around the snippet
+      const content = (element.content || '').trim();
+
+      return {
+        devtoolsNodePath: element.devtoolsNodePath,
+        content,
+        errors: await validateJsonLD(content),
+      };
+    }));
+    // Show invalid snippets at the top
+    validatedSnippets.sort((a, b) => {
+      return b.errors.length - a.errors.length;
+    });
+
+    const renderedSnippets = await Promise.all(
+      validatedSnippets.map(snippetWithErrors => renderValidatedSnippet(snippetWithErrors))
+    );
+    const details = Audit.makeListDetails(renderedSnippets);
+
+    const invalidSnippets = validatedSnippets.filter(vs => vs.errors.length > 0);
+    const validSnippets = validatedSnippets.filter(vs => vs.errors.length === 0);
+    const displayValue = str_(UIStrings.displayValue, {
+      validSnippetProportion: validSnippets.length / jsonLDElements.length,
+    });
+
+    return {
+      score: invalidSnippets.length === 0 ? 1 : 0,
+      details,
+      displayValue,
+    };
+  }
+}
+
+/**
+ * @param {{content: string, devtoolsNodePath: string, errors: LH.StructuredData.ValidationError[]}} validatedSnippet
+ */
+function renderValidatedSnippet(validatedSnippet) {
+  const {content, devtoolsNodePath, errors} = validatedSnippet;
+
+  let parsedContent;
+  let topLevelType;
+  let topLevelName;
+  try {
+    parsedContent = JSON.parse(content);
```

Comment:
can we reuse a type for this somewhere from sd-validation types?

Reply:
It's kinda like `ExpandedSchemaRepresentationItem` – is there actually anything special about that type being `Expanded`?

Reply:
I don't totally remember :) buuuuuuut I think it's just that there are more compact forms/shortcuts/references that get resolved to their actual values?

https://json-ld.org/spec/latest/json-ld-api/#expansion

You've been much deeper in this more recently than me though, so I'll leave it to your judgement. We are trying to access the `@type` and `name` properties though so have *some* type would be nice

Reply:
I've added a type for it now. Turns out we weren't actually returning a valid `ExpandedSchemaRepresentation` from `expand`, so fixed that type too.

> I don't totally remember :) buuuuuuut I think it's just that there are more compact forms/shortcuts/references that get resolved to their actual values?

That's a good point, ideally we should use the expanded representation and look for "http://schema.org/name" on there. What do you think of having the the validator return an object with `errors` and `expandedSchema` properties, instead of just the errors array?

Fwiw, I've never actually seen anyone do anything fancy with JSON-LD in the wild.

Reply:
Ah nice catch!! Yeah that sounds reasonable to me, though maybe if it's not particularly common we punt for now. Just file an issue?

Reply:
Filed: https://github.com/GoogleChrome/lighthouse/issues/8668